src/runtime/cuda/nccl_api.cu

```diff
@@ -0,0 +1,832 @@
+/*!
+ *  Copyright (c) 2021 by Contributors
+ * \file nccl_api.cc
+ * \brief Implementation of wrapper around NCCL routines.
+ */
+
+
+#include "nccl_api.h"
+
+#include <dgl/array.h>
+#include <dgl/aten/array_ops.h>
+#include <dgl/runtime/container.h>
+#include <dgl/runtime/device_api.h>
+#include <dgl/packed_func_ext.h>
+#include <dgl/runtime/registry.h>
+#include <cuda_runtime.h>
+#include <cuda_fp16.h>
+
+#include <cmath>
+#include <sstream>
+#include <iomanip>
+#include <utility>
+#include <vector>
+#include <memory>
+#include <string>
+#include <limits>
+
+#include "cuda_common.h"
+#include "../../partition/ndarray_partition.h"
+#include "../../kernel/cuda/atomic.cuh"
+#include "../../array/cuda/dgl_cub.cuh"
+#include "../../array/cuda/array_index_select.cuh"
+
+#define NCCL_CALL(func) \
+{ \
+  ncclResult_t result = func; \
+  if (result != ncclSuccess) { \
+      LOG(FATAL)                                                        \
+          << "NCCLError: " #func " failed with error: " << result;            \
+  } \
+}
+
+namespace dgl {
+
+using namespace kernel::cuda;
+using namespace partition;
+
+namespace runtime {
+namespace cuda {
+
+namespace {
+
+enum class AllToAllMode : int {
+  REMAINDER = 0
+};
+
+
+template<typename T> ncclDataType_t NCCLType();
+template<> ncclDataType_t NCCLType<int32_t>() {
+    return ncclInt32;
+}
+template<> ncclDataType_t NCCLType<int64_t>() {
+    return ncclInt64;
+}
+template<> ncclDataType_t NCCLType<__half>() {
+    return ncclHalf;
+}
+template<> ncclDataType_t NCCLType<float>() {
+    return ncclFloat32;
+}
+template<> ncclDataType_t NCCLType<double>() {
+    return ncclFloat64;
+}
+
+
+template<typename IdType> __global__ void _MapProcByRemainder(
```

Comment:
Same function as in src/partition/cuda/partition_op.cu.
Shall we remove this one?

Reply:
Removed in 
4a9a3c0